### PR TITLE
Assigment 6 Kasey Miller

### PR DESCRIPTION
**Pull request by 1Kmiller1**
Sorry For the Single commit I had to move the project back to assignment 6 after building it in the incorrect branch

### CODE REVIEW

**Review by tienv9 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ❌
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)
Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)

Fundamentals
Place all shared project properties into a Directory.Build.props file.
Place all shared project items into a Directory.Build.targets file. (optional)
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                program.WriteLine($"Solution : {result}");
+                break;
+            }
+            else
+            {
+                program.WriteLine("Invalid input.");
+            }
+        }while (true);
+    }
+
```
Comment (julc14) [inline Calculate/Program.cs:35]: nit: formatting

File: Calculator.tests/ProgramTests.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        
+        Console.SetIn(consoleReader);
+        Console.SetOut(consoleWriter);
+
+        // Act
+        Program.Main();
+
+        // Assert
+        var actual = consoleWriter.ToString();
+        Assert.Equal(expected, actual);
```
Comment (julc14) [inline Calculator.tests/ProgramTests.cs:32]: You should have at least 2 tests. one to validate the reader and one for the writer. Generally should have 1 assert per test

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public static double Divide(int i, int j)
+    {
+        if (j == 0)
+        {
+            throw new DivideByZeroException();
+        }
+        return (double)i / (double)j;
+    }
+
+    public IReadOnlyDictionary<char, Func<int, int, double>> ArithmeticOperators { get;} = new Dictionary<char, Func<int, int, double>>
```
Comment (julc14) [inline Calculate/Calculator.cs:24]: Property not named as per requirements

**Pull request closed without merging**